flutter: 3.3.3 -> 3.7.12, add Linux desktop support during development, improve wrapping, patching and determinism

hacker1024 opened the pull request:
###### Description of changes

This PR:
- Upgrades Flutter to 3.7.12, introducing support for the new dependency preloading mechanism used to ship dependencies of the Flutter tool in the SDK
- Takes control over engine artifact retrieval
  - Disables artifact updates, ensuring that the same artifacts shipped in the SDK release tarball are used by Flutter in every invocation
  - Removes dependencies on the FHS by patching prebuilt libraries and executables
- Improves `mkFlutterApp`
  - [Reduces (hopefully eliminates!) non-determinism](https://github.com/NixOS/nixpkgs/issues/211606#issuecomment-1521171576)
  - Refactors the dependency packaging to be reusable anywhere Dart is needed
  - Validates (and generates if missing) pubspec.lock files
  - Allows supplying custom pubspec.lock files
  - Splits debugging symbols into another output to reduce binary sizes
  - Adds a modular [package-based override system](https://github.com/NixOS/nixpkgs/pull/201027#issuecomment-1510437924) to allow Flutter packages to add dependencies and settings to application build environments
  - Renames it to `buildFlutterApplication` to fit in better with other framework derivation functions
- Improves target platform support
  - Moves the responsibility of providing Linux desktop dependencies from `mkFlutterApp` to the artifact derivations and wrapper, enabling building for Linux desktop during development
  - Removes the FHS for the Android SDK (Android SDK support should be and is implemented elsewhere, such as in [android-nixpkgs](https://github.com/tadfisher/android-nixpkgs))

Closes: #224090
Closes: #215754
Closes: #215537
Closes: #225627

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

@ericdallo @babariviere

hacker1024 committed:
flutter: Add dependencies for Linux desktop compilation

hacker1024 committed:
flutter: Move runtime libraries out of FHS using LD_LIBRARY_PATH

dotlambda commented on pkgs/development/compilers/flutter/flutter.nix:
```diff
@@ -3,35 +3,51 @@
 , patches
 , dart
 , src
+, supportsLinuxDesktop ? true
```
```suggestion
, supportsLinuxDesktop ? stdenv.isLinux
```

dotlambda commented:
I'm trying to package https://github.com/adil192/saber using this but get
```
building '/nix/store/v2qb5qjwb542irh3mgbjrls89i1yr6jf-saber-0.3.15-deps-flutter-v3.3.3-x86_64-linux.tar.gz.drv'...
unpacking sources
unpacking source archive /nix/store/xq35asfa8xjwcs3k7yhbwv0vgy1ckj8i-source
source root is source
patching sources
configuring
no configure script, doing nothing
installing
Setting "enable-linux-desktop" value to "true".

You may need to restart any open editors for them to read new settings.
Downloading Material fonts...                                      778ms
Downloading Gradle Wrapper...                                       26ms
Downloading package sky_engine...                                  258ms
Downloading flutter_patched_sdk tools...                           573ms
Downloading flutter_patched_sdk_product tools...                   575ms
Downloading linux-x64 tools...                                      4.5s
Downloading linux-x64/font-subset tools...                         221ms
Running "flutter pub get" in source...                             35.1s
Downloading linux-x64/linux-x64-flutter-gtk tools...             2,732ms
Downloading linux-x64-profile/linux-x64-flutter-gtk tools...      1,528ms
Downloading linux-x64-release/linux-x64-flutter-gtk tools...      1,415ms

💪 Building with sound null safety 💪

Building Linux application...
CMake is required for Linux development.
It is likely available from your distribution (e.g.: apt install cmake), or can be downloaded from https://cmake.org/download/
mv: cannot stat '.packages': No such file or directory
error: builder for '/nix/store/v2qb5qjwb542irh3mgbjrls89i1yr6jf-saber-0.3.15-deps-flutter-v3.3.3-x86_64-linux.tar.gz.drv' failed with exit code 1;
```

hacker1024 commented:
> 

How are you packaging it? The FHS env won't work alone, you'll need to add CMake + other build tools to the PATH, as is done in the `flutter` launch wrapper.

dotlambda commented:
> How are you packaging it?

https://github.com/NixOS/nixpkgs/pull/201075

hacker1024 commented:
Ah, `mkFlutterApp` uses the unwrapped version of Flutter for some reason. I'll see what I can do.

hacker1024 committed:
flutter: Use RPATH and CFLAGS to remove Linux desktop compilation and runtime FHS dependency

hacker1024 committed:
flutter: Modularize unwrapped, wrapped, and FHS components

hacker1024 commented:
I've redesigned the flutter package architecture. This will hopefully allow `mkFlutterApp` to be reimplemented, sharing much of the existing flutter package logic.

It would do something like this:

1. Use a wrapped, FHS Flutter package to run `flutter build linux`
2. Patch `libflutter_linux_gtk.so` to correct the RUNPATH due to its dependency on libepoxy

This new Flutter package now does the rest of the work in terms of providing CMake dependencies and library locations.

I'm curious to hear your thoughts on this @mkg20001.

hacker1024 committed:
flutter: Only enable Linux desktop support by default on Linux

hacker1024 committed:
flutter: Implement mkFlutter function using barebones FHS environment

hacker1024 committed:
flutter.mkFlutterApp: Use wrapped Flutter package during build

This ensures that all runtime dependencies are supplied.

hacker1024 committed:
flutter.mkFlutterApp: Don't use .packages

This file is deprecated, unused, and not generated in newer Flutter versions.

hacker1024 commented:
With these latest changes, `mkFlutterApp` works again. Some logic is still duplicated; there's no need for `mkFlutterApp` to include Flutter dependencies now, as far as I understand. I have no time to clean that up, though, and it shouldn't be much less efficient.

dotlambda commented:
> By downloading the Flutter SDK, you agree to the Google Terms of Service.

Something has to be marked as unfree.

mkg20001 reviewed (approved):
i like this a lot, it splits the logic neatly.

hacker1024 committed:
maintainers: add hacker1024

hacker1024 commented:
> > By downloading the Flutter SDK, you agree to the Google Terms of Service.
> 
> Something has to be marked as unfree.

The licensing is a little confusing. The repository includes a [BSD 3-Clause license](https://github.com/flutter/flutter/blob/master/LICENSE), and the [ToS that I believe the tool refers to](https://docs.flutter.dev/tos) simply applies a very lenient trademark to the Flutter branding. Does that really mean the package must be marked as unfree?

mkg20001 commented:
It may be more complicated, since the SDK is only required to build the package and I'm not sure if any unfree components get copied to the final built derivation.

mkg20001 commented on pkgs/development/compilers/flutter/wrapper.nix:
```diff
@@ -0,0 +1,90 @@
+{ lib
+, stdenv
+, flutter
+, supportsLinuxDesktop ? stdenv.isLinux
```
I think adding ` assert (stdenv.isLinux); after }: should do it

FlafyDev replied:
how about putting `platforms = [ "x86_64-linux" "aarch64-linux" ];` in the runCommand instead?

hacker1024 replied:
We don't even supply Dart sources for macOS. Perhaps we can indicate that Flutter in its entirety is Linux-only in a meta attribute?

hacker1024 replied:
Upon further inspection, none of the Darwin artifacts are included in the SDK release tarball either.

mkg20001 replied:
since flutter platforms is linux only the assertion would be redundant, so this is resolved

hacker1024 committed:
flutter: Add fake SDK derivation

hacker1024 commented:
I have added a new feature. Let me know if it's beyond the scope of this PR, but I think it's relevant enough.

Tools like the IntelliJ Flutter plugin require a complete Flutter SDK directory, and use the `flutter` binary inside it. In order to get these tools to use the wrapped copy of Flutter, this new derivation symlinks the unwrapped SDK, overwriting any relevant unwrapped files with their wrapped counterparts. The output directory in the Nix store serves as a usable Flutter SDK directory.

mkg20001 commented:
Seems good. Is there any way this can be handeled automatically so a derivation that uses flutter automatically uses fakeSdk when being called via nixShell? Perhaps a hook that is called in the shell that just appends the fake sdk at the beginning of the path if it detects that it's a nix shell environment.

hacker1024 commented:
I can look into it. The fake SDK isn't required for regular usage of the `flutter` tool, though, which is what most derivations/interactive CLI users would be doing with it. What would be the use case for such a hook?

mkg20001 commented:
So that when you want to develop on a project that has a flake or a nixpkgs derivation you can just run nix-shell and the IDE will be able to automatically use fakeSdk

teutat3s commented:
Sorry in advance for a bit of off-topic regarding this PR.

I'm trying to build the new `fluffychat` version `1.7.0` with these changes for the ZHF campaign, see my current WIP here:
https://github.com/NixOS/nixpkgs/compare/master...teutat3s:zhf/fluffychat

I hope to find advice how to fix the following error from you more experienced flutter nixpkgs maintainers:
```
CMake Error at /nix/store/rxb9zl1mdx9f2n3z0nyv9zhngj3f2pw0-cmake-3.24.3/share/cmake-3.24/Modules/FindPkgConfig.cmake:607 (message):
  A required package was not found                        ⣷
Call Stack (most recent call first):                      ⣯
  /nix/store/rxb9zl1mdx9f2n3z0nyv9zhngj3f2pw0-cmake-3.24.3/share/cmake-3.24/Modules/FindPkgConfig.cmake:829 (_pkg_check_modules_internal)
  flutter/ephemeral/.plugin_symlinks/flutter_secure_storage_linux/linux/CMakeLists.txt:13 (pkg_check_modules)
```

Here the relevant line from `CMakeLists.txt`:
```
──────────────────────────────────────────────────────────────────────
       │ File: /tmp/nix-build-fluffychat-1.7.0-deps-flutter-v3.3.8-x86_64-linux.tar.gz.drv-0/tmp.9iISRQ23k4/.pub-cache/hosted/pub.dartlang.org/flutter_secure_storage_linux-1.1.1/linux/CMakeLists.txt
...
  13   │ pkg_check_modules(LIBSECRET REQUIRED IMPORTED_TARGET libsecret-1>=0.18.4)
...
```

I have `libsecret` in the `buildInputs`, don't understand why it doesn't find it... does anybody have an idea?

hacker1024 commented:
> Sorry in advance for a bit of off-topic regarding this PR.
> 
> I'm trying to build the new `fluffychat` version `1.7.0` with these changes for the ZHF campaign, see my current WIP here: [master...teutat3s:zhf/fluffychat](https://github.com/NixOS/nixpkgs/compare/master...teutat3s:zhf/fluffychat)
> 
> I hope to find advice how to fix the following error from you more experienced flutter nixpkgs maintainers:
> 
> ```
> CMake Error at /nix/store/rxb9zl1mdx9f2n3z0nyv9zhngj3f2pw0-cmake-3.24.3/share/cmake-3.24/Modules/FindPkgConfig.cmake:607 (message):
>   A required package was not found                        ⣷
> Call Stack (most recent call first):                      ⣯
>   /nix/store/rxb9zl1mdx9f2n3z0nyv9zhngj3f2pw0-cmake-3.24.3/share/cmake-3.24/Modules/FindPkgConfig.cmake:829 (_pkg_check_modules_internal)
>   flutter/ephemeral/.plugin_symlinks/flutter_secure_storage_linux/linux/CMakeLists.txt:13 (pkg_check_modules)
> ```
> 
> Here the relevant line from `CMakeLists.txt`:
> 
> ```
> ──────────────────────────────────────────────────────────────────────
>        │ File: /tmp/nix-build-fluffychat-1.7.0-deps-flutter-v3.3.8-x86_64-linux.tar.gz.drv-0/tmp.9iISRQ23k4/.pub-cache/hosted/pub.dartlang.org/flutter_secure_storage_linux-1.1.1/linux/CMakeLists.txt
> ...
>   13   │ pkg_check_modules(LIBSECRET REQUIRED IMPORTED_TARGET libsecret-1>=0.18.4)
> ...
> ```
> 
> I have `libsecret` in the `buildInputs`, don't understand why it doesn't find it... does anybody have an idea?

Add `pkg-config` to the `nativeBuildInputs`. It has a hook that adds all included libraries to the search path.

teutat3s commented:
> Add `pkg-config` to the `nativeBuildInputs`. It has a hook that adds all included libraries to the search path.

I did that already, sadly the error stays the same :confused: https://github.com/NixOS/nixpkgs/compare/master...teutat3s:zhf/fluffychat#diff-b8a826e9eb66655fef059b3fde0f6dabaea8b1857d02a6c2b77bc9998605cfb5R48

hacker1024 committed:
flutter: Add libdeflate to build environment

hacker1024 committed:
flutter: Do not rely on pkg-config setup hook to build search path

hacker1024 committed:
flutter: Properly set linker flags in wrapper

hacker1024 committed:
flutter: Reorganize the wrapper code

hacker1024 committed:
flutter: Redesign wrapping architecture

The following principles are now in place:

- All wrappers will include SDK file symlinks. There is not much of a reason to not do so, and removing the option to omit it makes it easier to understand what each wrapper does.
- There is no longer a way to get the previous derivation from a wrapper. This could yield unexpected results based on the wrapping order. Instead, "sdk", "unwrapped", and "noFHS" passthru attributes are provided where appropriate.

algorithmiker committed:
maintainers: add FlafyDev

hacker1024 commented:
@gilice I'd like to draw your attention to this as you've been doing some great Flutter work. What do you think?

I _think_ these changes are compatible with #224090, but I see you've changed the FHS a bit so that may need updating here.

hacker1024 commented:
> Hey! I will take a look. In the meantime, do you maybe have a Matrix account?

@gilice Yep, I've never used it for DMs before but I think I've set it up right. My account is @hacker1024:matrix.org and my identity server is https://scalar.vector.im.

hacker1024 committed:
flutter: Create the flutter derivation with callPackage alone

hacker1024 committed:
flutter: Add cache output and preload pub cache when necessary

Co-authored-by: FlafyDev <flafyarazi@gmail.com>
Co-authored-by: gilice <gilice@proton.me>
Co-authored-by: hacker1024 <hacker1024@users.sourceforge.net>

hacker1024 committed:
flutter: Don't delete SDK artifacts

Co-authored-by: FlafyDev <flafyarazi@gmail.com>
Co-authored-by: gilice <gilice@proton.me>
Co-authored-by: hacker1024 <hacker1024@users.sourceforge.net>

hacker1024 committed:
flutter: Use autoPatchelfHook for SDK artifacts

This allows SDK artifacts (including build tools and libraries) to use the right interpreter and library locations, abolishing the need for a FHS or post-build patches.

`autoPatchelf` has been used through `autoPatchelfHook`, instead of the `autoPatchelf` command on just the artifact directory, due to the latter method not patching interpreter paths (https://discourse.nixos.org/t/trying-to-use-autopatchelf/24003).

Co-authored-by: FlafyDev <flafyarazi@gmail.com>
Co-authored-by: gilice <gilice@proton.me>
Co-authored-by: hacker1024 <hacker1024@users.sourceforge.net>

hacker1024 committed:
flutter: Add zlib to the linker search path

hacker1024 committed:
flutter: Remove the FHS wrapper

Now that Flutter artifacts are patched, there is no need for a FHS at runtime or any stage of the build process.

Those relying on it for the Android SDK should look at https://github.com/tadfisher/android-nixpkgs.

hacker1024 committed:
flutter: Don't use an external artifact cache

The internal artifacts should be used, as they have been patched to use Nix libraries.

Co-authored-by: FlafyDev <flafyarazi@gmail.com>
Co-authored-by: gilice <gilice@proton.me>
Co-authored-by: hacker1024 <hacker1024@users.sourceforge.net>

hacker1024 committed:
flutter: Don't try to update the immutable artifact cache

Co-authored-by: FlafyDev <flafyarazi@gmail.com>
Co-authored-by: gilice <gilice@proton.me>
Co-authored-by: hacker1024 <hacker1024@users.sourceforge.net>

hacker1024 committed:
flutter: Don't preserve prebuilt SDK asset file permissions

See: https://github.com/flutter/flutter/issues/119961

Co-authored-by: FlafyDev <flafyarazi@gmail.com>
Co-authored-by: gilice <gilice@proton.me>
Co-authored-by: hacker1024 <hacker1024@users.sourceforge.net>

hacker1024 committed:
flutter: Remove prebuilt dependencies from the wrapper

These are no longer needed now that the artifacts are patched.

hacker1024 committed:
flutter: 3.3.3 -> 3.7.11

hacker1024 committed:
flutter: Don't check executable locations in doctor

Flutter 3.5.0 adds a check to make sure that the Flutter and Dart tool executables are inside the SDK directory.

This does not play nicely with the wrapping and linking techniques used to package it. It is also impossible for there to be an issue related to binary locations, as everything is set up properly in the derivation.

hacker1024 committed:
Merge remote-tracking branch 'NixOS/master' into patch-3

hacker1024 committed:
flutter: Remove stray dollar sign from name string

hacker1024 committed:
flutter: Add FlafyDev, gilice, and hacker1024 as maintainers

hacker1024 committed:
flutter: Add git and which to PATH

hacker1024 committed:
mkFlutterApp: Deduplicate build environment setup

The build environment is now set up by the Flutter wrapper, and does not need to be done in mkFlutterApp.

hacker1024 committed:
mkFlutterApp: Don't fetch prebuilt artifacts

Artifacts are now supplied by the Flutter SDK derivation.

In addition, they are no longer a factor in the dependency hash calculation.

hacker1024 committed:
flutter: Add mkFlutterApp as a passthru attribute

hacker1024 commented:
I've got mkFlutterApp working. It's not longer in charge of configuring the build environment or supplying general build dependencies - this was re-implemented in the wrapper earlier. It also no longer fetches artifacts or includes them in hash calculations.

Can anyone confirm the yubioath-flutter vendor hash on x64? I get sha256-JQY4Uv0K3OFaiJy2k31TUTKk62PubjWdzGo9TukYVkE=.

hacker1024 committed:
flutter: Move makeWrapper to nativeBuildInputs in wrapper

hacker1024 committed:
flutter: Rename "self" to "unwrapped" in main derivation

hacker1024 commented on pkgs/development/compilers/flutter/flutter.nix:
```diff
@@ -1,190 +1,131 @@
+    libepoxy
+  ];
+
+  self =
```
Done

hacker1024 committed:
flutter.mkFlutterApp: Don't use autoPatchelf on package cache

While this may be beneficial for packages that ship native binaries, this causes issues due to Flutter packages that have inappropriately uploaded build artifacts (such as those from example subprojects) to pub.dev.

hacker1024 committed:
flutter.mkFlutterApp: Don't vendor arbitrary configuration files

hacker1024 commented:
New problem: Linux ARM artifacts do not ship in the SDK release archive.
Is there a way to grab them manually in a stable manner? If not, we may need to drop ARM support.

hacker1024 commented on pkgs/build-support/flutter/default.nix:
```diff
@@ -1,36 +1,10 @@
-, xz
+, llvmPackages_13
+, cacert
+, flutter
```
I'd rather not force-push just for this. If there's a really compelling reason to reduce the diff to this degree then maybe we can before merging.

hacker1024 commented:
Something that may be worth looking into is if the artifacts are actually stable respective to the engine version. The [URL](https://github.com/flutter/flutter/blob/3.7.11/packages/flutter_tools/lib/src/cache.dart#L894) includes the version string.

Our non-determinism so far has come from hashing the artifacts, but we could maintain a list of hashes for each artifact, updating it with each engine update, and then package each individual artifact in its own fixed-output derivation.

hacker1024 committed:
flutter.mkFlutterApp: Add Git to nativeBuildInputs

The deterministic Git script depends on it.

mkg20001 commented on pkgs/development/compilers/flutter/packages/libdeflate/default.nix:
```diff
@@ -0,0 +1,17 @@
+{ stdenv, fetchFromGitHub, cmake }:
+
+stdenv.mkDerivation rec {
+  pname = "libdeflate";
```
libdeflate is already packaged in pkg/development/libraries/libdeflate/default.nix with the same version, please use or patch that if you can (it doesn't have outputs specified so you could patch that in)

hacker1024 replied:
The one in nixpkgs is very old and uses a different build system. It should probably be updated, but I figured that was beyond the scope of this PR.

hacker1024 replied:
Ah my bad, it has been updated. I'll use it.

mkg20001 reviewed (commented):
nice work :tada:

FlafyDev commented:
> New problem: Linux ARM engine artifacts do not ship in the SDK release archive. Is there a way to grab them manually in a reproducible manner? If not, we may need to drop ARM support until #212328 is complete.

`ec975089acb540fc60752606a3d3ba809dd1528b` is the engine version for the flutter 3.7.9. you can get it from `bin/internal/engine.version`

The command `gsutil ls -r gs://flutter_infra_release/flutter/ec975089acb540fc60752606a3d3ba809dd1528b`

returns the following:

```
...

gs://flutter_infra_release/flutter/ec975089acb540fc60752606a3d3ba809dd1528b/linux-arm64-debug/:
gs://flutter_infra_release/flutter/ec975089acb540fc60752606a3d3ba809dd1528b/linux-arm64-debug/linux-arm64-flutter-gtk.zip

gs://flutter_infra_release/flutter/ec975089acb540fc60752606a3d3ba809dd1528b/linux-arm64-profile/:
gs://flutter_infra_release/flutter/ec975089acb540fc60752606a3d3ba809dd1528b/linux-arm64-profile/linux-arm64-flutter-gtk.zip

gs://flutter_infra_release/flutter/ec975089acb540fc60752606a3d3ba809dd1528b/linux-arm64-release/:
gs://flutter_infra_release/flutter/ec975089acb540fc60752606a3d3ba809dd1528b/linux-arm64-release/linux-arm64-flutter-gtk.zip

gs://flutter_infra_release/flutter/ec975089acb540fc60752606a3d3ba809dd1528b/linux-arm64/:
gs://flutter_infra_release/flutter/ec975089acb540fc60752606a3d3ba809dd1528b/linux-arm64/artifacts.zip
gs://flutter_infra_release/flutter/ec975089acb540fc60752606a3d3ba809dd1528b/linux-arm64/font-subset.zip
gs://flutter_infra_release/flutter/ec975089acb540fc60752606a3d3ba809dd1528b/linux-arm64/linux-arm64-flutter-gtk.zip

...
```

I think this means we have to download each of the zip files and extract them in the `bin/cache/artifacts/engine/xx` directory where xx is the folder before the `:`

It doesn't seem `linux-arm64-debug` is required to be downloaded.

We probably want to download all of these in the same FOD derivation instead of fetchzip individually, otherwise there would be a lot of hashes to change for every flutter update. 

To download you change the gs:// to https://storage.googleapis.com/
like this: https://storage.googleapis.com/flutter_infra_release/flutter/ec975089acb540fc60752606a3d3ba809dd1528b/linux-arm64/linux-arm64-flutter-gtk.zip

hacker1024 committed:
flutter: Use existing libdeflate derivation

hacker1024 committed:
flutter.mkFlutterApp: Remove the target architecture from the dependency derivation name

As the dependency tarballs no longer include SDK artifacts, they are not platform dependent.

hacker1024 commented:
The reason I was thinking about very granular derivations was to allow users to save space. Artifacts that are not needed (such as if the user does not care for desktop support) could be left out with a package override without breaking any hashes.

Updating could be automated with a script.

Thoughts?

FlafyDev commented:
> The reason I was thinking about very granular derivations was to allow users to save space. Artifacts that are not needed (such as if the user does not care for desktop support) could be left out with a package override without breaking any hashes.
> 
> Updating could be automated with a script.
> 
> Thoughts?

if it can be done with a script then yeah it could be worth doing it that way..

also, what do you consider an artifact? for example, will all of the files required for linux-arm64 be a single derivation, or would `linux-arm64`, `linux-arm64-release`, `linux-arm64-profile` be separate derivations that will all be added to flutter if the user enables building to linux arm?

I think the latter makes more sense

hacker1024 commented:
I think it makes sense to split up debug, profile, and release derivations. Developers will typically want all three, but `mkFlutterApp` only needs release.

We can be clever with `symlinkJoin` too to avoid rebuilding the entire `flutter` derivation for different artifact combinations.

hacker1024 committed:
flutter: Remove --no-version-check from the immutable wrapper

This is not technically needed anyway, as the feature is removed in disable-auto-update.patch.

Prepending this argument causes issues:
- https://github.com/kevmoo/completion.dart/issues/33
- https://github.com/flutter/flutter/issues/124793

hacker1024 committed:
flutter: Use fetchzip instead of fetchurl for source downloads

This allows for speedier re-use of the source archives, and lets files inside be read in Nix.

hacker1024 committed:
flutter: Download engine artifacts individually

This brings the following benefits:
- Artifacts missing in the SDK tarball (such as prebuilts for linux-aarch64) can be obtained
- Artifacts can be patched more granularly (e.g. libflutter_linux_gtk is patchelf-ed for GTK3, and the linux-aarch64 assets have a postPatch to add some missing files)
- Minimal Flutter packages can be generated (e.g. mkFlutterApp only needs prebuilts for desktop Linux, and developers that don't care for desktop Linux can exclude the artifacts and dependencies)

It also paves the way for including manually built engine facts.

hacker1024 commented:
I have implemented granular artifact support. Unfortunately, `symlinkJoin` cannot be used as Flutter requires artifacts to be in the same SDK directory as the real `flutter` executable.

This brings the following benefits:
- Artifacts missing in the SDK tarball (such as prebuilts for linux-aarch64) can be obtained
- Artifacts can be patched more granularly (e.g. libflutter_linux_gtk is patchelf-ed for GTK3, and the linux-aarch64 assets have a postPatch to add some missing files)
- Minimal Flutter packages can be generated (e.g. `mkFlutterApp` only needs prebuilts for desktop Linux, and developers that don't care for desktop Linux can exclude the artifacts and dependencies)

To make the updater script, something along the lines of `nix-instantiate --eval --expr 'builtins.toJSON (import ./hashes.nix)'` can be done, where the resultant JSON keys can be turned into download URLs.

hacker1024 committed:
flutter.mkFlutterApp: Don't hash more unstable files

Files and directories such as .dart_tool, .flutter-plugins, .packages and .pub-cache/hosted/*/.cache will be happily regenerated by the SDK in an offline environment.

The patches being made to these files and directories to remove non-determinism were flawed: They did not handle cases where files only appeared in one run, or cases where tooling versions had an effect on the output.

hacker1024 commented on pkgs/development/compilers/flutter/engine-artifacts/default.nix:
```diff
@@ -0,0 +1,127 @@
+      };
+    };
+
+  mkArtifactDerivation = { platform ? null, variant ? null, archive, ... }@args:
```
Some artifacts are not unique to any platforms or build variants (see [here](https://github.com/NixOS/nixpkgs/blob/62e50080f39560183c57aa184bf03e47a341444c/pkgs/development/compilers/flutter/engine-artifacts/default.nix#L103)).

hacker1024 committed:
flutter: Pass through the version attribute in the wrapper

hacker1024 committed:
flutter: Don't tarball the package cache

Co-authored-by: FlafyDev <flafyarazi@gmail.com>

The dependency fixed-output derivation now uses the recursive hash mode to avoid tarballing and copying all the files.

https://github.com/NixOS/nix/issues/6660 was ran into during the development of this change. Input references were found in the Git package cache before nukeReferences was used.

It turns out that the mirrors in the Git package cache do not actually need to be preserved, as the SDK does not use them during the build process. They are therefore deleted in the dependency derivation and re-created as blank repositories in the main derivation.

A description of the Git package cache layout can be found here: https://github.com/dart-lang/pub/blob/c890afa1d65b340fa59308172029680c2f8b0fc6/lib/src/source/git.dart#L339

hacker1024 committed:
flutter: Remove cached Git package Git directories

Pub does not perform any Git operations on the cached package directly, instead cloning it through a mirror.

The .git directories are not needed, and are a source of non-determinism.

hacker1024 committed:
flutter.mkFlutterApp: Rename to flutter.buildFlutterApplication

This fits more appropriately with other build derivations like buildPythonApplication.

hacker1024 committed:
flutter.buildFlutterApplication: Refactor dependency setup as a standalone derivation with an output hook

This allows anything that needs to build a Dart application (such as a future `buildDartApplication` function) to use the same setup.

algorithmiker committed:
flutter.buildFlutterApplication: Add an explanation when a pubspec mismatch occurs

hacker1024 committed:
dart.fetch-dart-deps: Remove stray comment from the dependency derivation

FlafyDev commented on pkgs/build-support/dart/fetch-dart-deps/default.nix:
```diff
@@ -0,0 +1,133 @@
+      outputHashAlgo = "sha256";
+      outputHashMode = "recursive";
+      outputHash = if vendorHash != "" then vendorHash else lib.fakeSha256;
+    } // drvArgs);
```
Wouldn't this override the `name` declared at line 59?

hacker1024 committed:
flutter.buildFlutterApplication: Allow using a custom pubspec.lock

hacker1024 committed:
dart.fetch-dart-deps: Don't delete .git/pub-packages

hacker1024 committed:
yubioauth-flutter: Update vendor hash for new pubspec.lock handling

hacker1024 committed:
firmware-updater: Add a pubspec.lock

hacker1024 committed:
dart.fetch-dart-deps: Verify the pubspec.lock as well as pubspec.yaml

88275ca6d6ceeaecfc9c7d1a313c198bb888d287 inadvertently stopped the pubspec.lock from changing, as it copied the file before running pub get.

pub get can modify the pubspec.lock to update it to newer formats (for example by adding hashes and updating URLs, see an example diff below). We do not need the modifications at any later stage, so we can preserve the original file.

      boolean_selector:
        dependency: transitive
        description:
          name: boolean_selector
    +      sha256: "6cfb5af12253eaf2b368f07bacc5a80d1301a071c73360d746b7f2e32d762c66"
    +      url: "https://pub.dev"
    -      url: "https://pub.dartlang.org"
        source: hosted
    +    version: "2.1.1"
    -    version: "2.1.0"

hacker1024 committed:
dart.fetch-dart-deps: Don't allow overriding the deps derivation name

hacker1024 committed:
dart.fetch-dart-deps: Compress the generated pubspec.lock file

This is much more practical for users to copy, as it significantly reduces the length of the output.

hacker1024 committed:
dart.fetch-dart-deps: Fix invalid syntax in pubspec validation statement

hacker1024 commented:
It's common for Dart packages and especially Flutter plugins to require some additional patching for Nix. Take [handy_window](https://pub.dev/packages/handy_window), for example, which is unable to find C/C++ build dependencies without modification.

I think it might be worthwhile to put packages into derivations to let them be easily patched and overridden. Something along these lines:

- A standard `prepareDartPackage` derivation function is added to Nixpkgs. This prepares package sources and applies patches that are useful to all Dart packages.
- A map of package names to customized derivation functions that extend `prepareDartPackage` is added.
- A new derivation looks through the `fetch-dart-deps` output. It:
  - Checks if each cached package exists in the aforementioned map.
  - Creates a derivation (using the generic `prepareDartPackage` function if none other is specified), using the cached package directory as the `src`.
  - Links the derivation output to same place the original cached package was found.

This way, most packages will work well without intervention, but those that require modification can be fixed.

What do you think?

FlafyDev commented:
> It's common for Dart packages and especially Flutter plugins to require some additional patching for Nix. Take [handy_window](https://pub.dev/packages/handy_window), for example, which is unable to find C/C++ build dependencies without modification.
> 
> I think it might be worthwhile to put packages into derivations to let them be easily patched and overridden. Something along these lines:
> 
>     * A standard `prepareDartPackage` derivation function is added to Nixpkgs. This prepares package sources and applies patches that are useful to all Dart packages.
> 
>     * A map of package names to customized derivation functions that extend `prepareDartPackage` is added.
> 
>     * A new derivation looks through the `fetch-dart-deps` output. It:
>       
>       * Checks if each cached package exists in the aforementioned map.
>       * Creates a derivation (using the generic `prepareDartPackage` function if none other is specified), using the cached package directory as the `src`.
>       * Links the derivation output to same place the original cached package was found.
> 
> 
> This way, most packages will work well without intervention, but those that require modification can be fixed.
> 
> What do you think?

So this essentially allows you to edit each package? Wouldn't it be preferred to do that in the preBuild stage that a specific app, like fluffychat, writes?

hacker1024 commented:
> > It's common for Dart packages and especially Flutter plugins to require some additional patching for Nix. Take [handy_window](https://pub.dev/packages/handy_window), for example, which is unable to find C/C++ build dependencies without modification.
> > I think it might be worthwhile to put packages into derivations to let them be easily patched and overridden. Something along these lines:
> > ```
> > * A standard `prepareDartPackage` derivation function is added to Nixpkgs. This prepares package sources and applies patches that are useful to all Dart packages.
> > 
> > * A map of package names to customized derivation functions that extend `prepareDartPackage` is added.
> > 
> > * A new derivation looks through the `fetch-dart-deps` output. It:
> >   
> >   * Checks if each cached package exists in the aforementioned map.
> >   * Creates a derivation (using the generic `prepareDartPackage` function if none other is specified), using the cached package directory as the `src`.
> >   * Links the derivation output to same place the original cached package was found.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This way, most packages will work well without intervention, but those that require modification can be fixed.
> > What do you think?
> 
> So this essentially allows you to edit each package? Wouldn't it be preferred to do that in the preBuild stage that a specific app, like fluffychat, writes?

If it's done in application derivations, it become hard to share fixes for common dependencies between applications.

hacker1024 committed:
flutter.buildFlutterApplication: Add flutterBuildFlags argument

hacker1024 committed:
flutter.buildFlutterApplication: Move debugging symbols into another output

This reduces the size of the executable.

It works well with Nix - if the user includes the version in their bug report, the exact debugging symbols required can easily be found.

hacker1024 commented:
I have made a change to move debugging symbols into a separate output. This has very significant results: The `firmware-updater` executable is now ~ 24 KB, while the debugging symbols are ~2.9 MB.

hacker1024 commented:
> It's common for Dart packages and especially Flutter plugins to require some additional patching for Nix. Take [handy_window](https://pub.dev/packages/handy_window), for example, which is unable to find C/C++ build dependencies without modification.
> 
> I think it might be worthwhile to put packages into derivations to let them be easily patched and overridden. Something along these lines:
> 
>     * A standard `prepareDartPackage` derivation function is added to Nixpkgs. This prepares package sources and applies patches that are useful to all Dart packages.
> 
>     * A map of package names to customized derivation functions that extend `prepareDartPackage` is added.
> 
>     * A new derivation looks through the `fetch-dart-deps` output. It:
>       
>       * Checks if each cached package exists in the aforementioned map.
>       * Creates a derivation (using the generic `prepareDartPackage` function if none other is specified), using the cached package directory as the `src`.
>       * Links the derivation output to same place the original cached package was found.
> 
> 
> This way, most packages will work well without intervention, but those that require modification can be fixed.
> 
> What do you think?

Implementing this is tricky. `handy_window`, for example, requires some include directories to be added to `CFLAGS`.

```dart
{ flutter, fribidi, cairo }:

flutter.buildFlutterApplication rec {
  # ...
  CFLAGS = "-isystem ${fribidi}/include/fribidi -isystem ${cairo.dev}/include";
}
```

I think the easiest way to implement this using the package derivation system I described is with a setup hook that adds to the `CFLAGS` environment variable. The issue here is making sure that the `target` version of the dependencies are used.

A naive approach in the setup hook would be to use substitutions to obtain the include paths:

```nix
{ prepareDartPackage, fribidi, cairo }:
src:

prepareDartPackage {
  inherit src;
  CFLAGS = "-isystem ${fribidi}/include/fribidi -isystem ${cairo.dev}/include";
  setupHook = ./setup-hook.sh;
}
```

```
preConfigureHooks+=(_setupHandyWindowBuildEnv)

_setupHandyWindowBuildEnv() {
  export CFLAGS="$CFLAGS @CFLAGS@"
}
```

This would be problematic in a cross-compiling setup, though, as the substitutions used to create the setup hook would be using host packages, not target packages.

How can we make this work taking the target into account?

The [cc-wrapper](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/cc-wrapper/setup-hook.sh) and [pkg-config-wrapper](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/pkg-config-wrapper/setup-hook.sh) deal with a similar situation, but they don't have to deal with paths that are unique to each package. The `fribidi` include here, for example, uses `include/fribidi`, while the `cairo` include just uses `include`.

hacker1024 commented:
The other way to go about doing this, which is a lot simpler, is to use passthru attributes to configure the build environment. E.g.

```
{ prepareDartPackage, fribidi, cairo }:
src:

prepareDartPackage {
  inherit src;
  passthru.includeDirectories = [
    (fribidi + /include/fribidi)
    (cairo.dev + /include)
  ];
}
```

This is less flexible, though.

hacker1024 commented:
Maybe package derivations are the wrong approach. What if instead we just had a map of package names to attribute sets that contain attributes to add to the `buildFlutterApplication` call?

This is similar to how Home Assistant on NixOS works - it has a map of component names to dependencies to add.

hacker1024 committed:
flutter.buildFlutterApplication: Supply runtime dependencies

This wraps Flutter programs with an appropriate LD_LIBRARY_PATH.

For some reason, the RUNPATH of the executable is not used to load dynamic libraries in dart:ffi with DynamicLibrary.open().

This could alternatively be fixed with patchelf --add-needed, but this would cause all the libraries to be opened immediately,
which is not what application authors expect.

The name of the runtimeDependencies argument was chosen to match autoPatchelfHook, which has a similar feature.

hacker1024 committed:
flutter: Allow adding extra dependencies to the wrapper

Flutter application derivations can add dependencies and build flags, but it's not as easy during development.

As an alternative to nix-shell, a custom Flutter wrapper can be built with dependencies required by projects.

This feature builds on the existing build environment setup techniques required for Linux desktop support, so is fairly cheap to maintain.

hacker1024 committed:
dart.fetch-dart-deps: Expose the package cache files through passthru

hacker1024 committed:
dart.fetch-dart-deps: Add hook runtime dependencies

hacker1024 committed:
dart.list-dart-deps: Add list-dart-deps function

This uses `dart pub deps` (https://dart.dev/tools/pub/cmd/pub-deps) to retrieve information about project dependencies.

It requires a fetch-dart-deps derivation as input.

hacker1024 committed:
flutter.buildFlutterApplication: Allow overriding configuration attributes

This allows thing like `flutterBuildFlags` and `runtimeDependencies` to be overridden with `overrideAttrs`.

Attributes that affect dependency retrieval cannot be overridden.

hacker1024 committed:
flutter.buildFlutterApplication: Introduce a package override repository

Some Flutter packages require additional attribute values to be added to buildFlutterApplication, to add things like libraries and environment variables.

To prevent duplication in applications that use the packages, a repository of package overrides is kept. buildFlutterApplication will look for package overrides for each dependency, and apply them by calling overrideAttrs on itself.

hacker1024 commented:
I have added some features based on the ideas above. Please see an example of using them [here](https://github.com/NixOS/nixpkgs/compare/0495725a1f8e9df1ac3996459a9bb82305e69242...2d417dc283977d90873dd5215c7cea155ca919b1), where I've converted FluffyChat back into a source-based build.

Interestingly, none of the native software that FluffyChat depends on are direct dependencies; rather, they are dependencies of Flutter plugins that FluffyChat uses. This will be the case for the majority of Flutter applications.

Now, the package dependencies are tied to the packages themselves, which means Flutter application derivations do not need to worry about them.

hacker1024 commented on pkgs/build-support/dart/fetch-dart-deps/default.nix:
```diff
@@ -0,0 +1,163 @@
+                       '\n     https://dart.dev/guides/libraries/private-files#pubspeclock' \
+                       '\n  2. An attempt to generate and print a compressed pubspec.lock file will be made now.' \
+                       '\n     It is compressed with gzip and base64 encoded.' \
+                       '\n     Paste it to a file and extract it with `base64 -d pubspec.lock.in | gzip -d > pubspec.lock`.' \
```
See: https://github.com/NixOS/nixpkgs/pull/201027/commits/075f50f2d167affaefe3783dc21922a8288f4247

hacker1024 commented on pkgs/build-support/dart/fetch-dart-deps/default.nix:
```diff
@@ -0,0 +1,163 @@
+                       '\n     This must be updated whenever the application is updated.' \
+                       '\n'
+          _pub_get
+          echo ""
```
This adds a newline after calling `pub get` so the pubspec.lock is easily visible.

hacker1024 commented on pkgs/build-support/dart/list-dart-deps/default.nix:
```diff
@@ -0,0 +1,28 @@
+}:
+deps:
+
+builtins.fromJSON (builtins.readFile (stdenv.mkDerivation {
```
It is used [here](https://github.com/NixOS/nixpkgs/blob/0495725a1f8e9df1ac3996459a9bb82305e69242/pkgs/build-support/flutter/default.nix#L112).

It is IFD, but it will generally only be evaluated right before a package build so I think it's all right. There's no easy way to do this without IFD.

hacker1024 committed:
flutter.buildFlutterApplication: Supply package metadata to overrides

hacker1024 commented:
I have updated the package override system to make package metadata available. This allows overrides to take things like the package version, source (e.g. `root`, `hosted`, `path`, `git`) and dependency type (e.g. `root`, `direct`, `transitive`) into consideration.

hacker1024 commented on pkgs/build-support/flutter/default.nix:
```diff
@@ -115,7 +115,16 @@ in
+  then
+    prev.overrideAttrs
+      (packageOverrideRepository.${package.name} {
+        inherit (package)
```
The whole `package` attribute is actually inherited here, but I list the attribute names explicitly for a couple of reasons.

1. If the format of `dart pub deps --json` changes (which is unlikely as the `dart pub deps` output doesn't change much in general, and `--json` hasn't changed since its initial implementation in March 2021), evaluation will fail once in a consistent place
2. It makes the available metadata attributes clear

hacker1024 commented:
> Hydra cannot IFD so I don't think this is allowed in nixpkgs.

Well, that's disappointing. To avoid manually maintaining a list of dependencies in each application, I propose this solution:

Add a `depList` argument in `buildFlutterApplication` that should contain the contents of `dart pub deps --json`. The derivation used in `fetch-dart-deps` can be added as a passthru attribute in `buildFlutterApplication`.

The workflow for packaging a Flutter application would look like this:

1. Write the derivation with `buildFlutterApplication` (e.g. called `myFlutterDerivation`).
2. Build the `myFlutterDerivation.depList` derivation. This will generate a JSON file.
3. Add the JSON file to the `buildFlutterApplication` call's `depList` argument (via `builtins.fromJSON`)
4. Attempt to build `myFlutterDerivation`. If a `pubspec.lock` needed to be generated, save it and use it in the `buildFlutterApplication` call's `pubspecLockFile` argument.
6. Attempt to build and run `myFlutterDerivation`. If any errors occur, add the required changes to either the `myFlutterDerivation` or a package override.

To ensure that the `depsList` does not get forgotten with updates, it could contain the `vendorHash`, which would be validated.

Does anyone have any issues with this?

hacker1024 committed:
flutter.buildFlutterApplication: Manually supply the dependency list

This removes the need for IFD.

hacker1024 committed:
flutter.buildFlutterApplication: Add an option to generate the dependency list with IFD

This is useful for projects outside of Nixpkgs.

hacker1024 commented:
I have implemented the design from my previous comment. I have also preserved the IFD functionality through the `autoDepsList` option, which is convenient for projects outside of Nixpkgs.

hacker1024 committed:
flutter: 3.7.11 -> 3.7.12

hacker1024 committed:
flutter: Throw a useful message when there are missing artifact hashes

hacker1024 committed:
flutter: Don't use IFD to read the engine version

hacker1024 committed:
Merge remote-tracking branch 'NixOS/master' into feature/futter-linux-desktop

hacker1024 committed:
flutter: Generate target PKG_CONFIG_PATH at build time

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2130

mkg20001 commented on pkgs/development/compilers/flutter/package-overrides/default.nix:
```diff
@@ -0,0 +1,4 @@
+{ callPackage }:
+
+{
```
is this going to be used? is it just intentionally left empty as we do not need any patches right now?

hacker1024 replied:
It [will be used](https://github.com/NixOS/nixpkgs/pull/201027#issuecomment-1510437924), but that can come in another PR.

mkg20001 reviewed (approved):
it's a lot better than the existing implementation, so I'd give it a go

I didn't review the whole in detail but overall it appears sane

mkg20001 commented:
in the future (that means different pr, since this one is big enough already) a script for setting up (or updating) the two lockfiles might be cool.
perhaps a generic updater `flutter.mkUpdateScript "<package>";`

mkg20001 commented:
@hacker1024 you could pull request the fluffychat changes if you want now

hacker1024 commented:
> in the future (that means different pr, since this one is big enough already) a script for setting up (or updating) the two lockfiles might be cool. perhaps a generic updater `flutter.mkUpdateScript "<package>";`

The `buildFlutterApplication` derivation will generate and print these if they're missing. I could build an update script pretty easily but I figured that's simple enough.

mkg20001 commented:
@kirillrdy @hacker1024  It appears that it's already cached on cache.nixos.org, building without cache fails indeed. Didn't check that.

This would need fixing. Also does this mean that the assets under a google storage url get changed without notice?

hacker1024 commented:
> @kirillrdy @hacker1024 It appears that it's already cached on cache.nixos.org, building without cache fails indeed. Didn't check that.
> 
> This would need fixing. Also does this mean that the assets under a google storage url get changed without notice?

I'll run a diff soon and take a look at what changed. If this is the case, we may need to find another way to get assets in a stable manner. If there's no obvious answer we'll need to talk to Google about it.

Once the engine PR is ready this will no longer be an issue.

hacker1024 commented:
> this fails to build using merge commit [9096817](https://github.com/NixOS/nixpkgs/commit/9096817104bae9d7bfa214c4486ef8c6df6bf34b)
> 
> ```
> nix-build -A flutter
> these 20 derivations will be built:
>   /nix/store/47m7nl28jk3nh47mf36b2807gx50g4sr-source.drv
>   /nix/store/yspn0q357wzbzmw2pa9l19194dyb8zj5-source.drv
>   /nix/store/8v8a0cpkk5dgaw0cnsxwg32zi0mfrii1-flutter-artifact-android-x86-jit-release-artifacts-1a65d409c7a1438a34d21b60bf30a6fd5db59314.drv
>   /nix/store/yqd15cbi655kj6ysx5n1bjn1y71lydc5-source.drv
>   /nix/store/cyhfnbm1v2gv4k864kjfqzxnba94fmrb-flutter-artifact-android-arm64-release-linux-x64-1a65d409c7a1438a34d21b60bf30a6fd5db59314.drv
>   /nix/store/y0kyyb5h946xpdw51k9g3lw3wf187yl9-source.drv
>   /nix/store/dvfiwhfq7g0azpywqxbf8dcnn0zmr03s-flutter-artifact-android-arm-artifacts-1a65d409c7a1438a34d21b60bf30a6fd5db59314.drv
>   /nix/store/61grn00z09rs758ds8x01vgvlzv38s4x-source.drv
>   /nix/store/hk65klcklshkfdqjh0pb9bgrvm81jsar-flutter-artifact-linux-x64-release-linux-x64-flutter-gtk-1a65d409c7a1438a34d21b60bf30a6fd5db59314.drv
>   /nix/store/5mvrsprslvr0fg225xymjvyk15il8g45-source.drv
>   /nix/store/j6fgf61bpndxi0ivphdyh5hx6w117h61-flutter-artifact-linux-x64-profile-linux-x64-flutter-gtk-1a65d409c7a1438a34d21b60bf30a6fd5db59314.drv
>   /nix/store/zaja9cgsn70ipdap8aknrb3w46fwdxnx-source.drv
>   /nix/store/s7mdfiy6q45zysx88f7xmcw4qy00szxc-flutter-artifact-android-x64-release-artifacts-1a65d409c7a1438a34d21b60bf30a6fd5db59314.drv
>   /nix/store/498dcxxl3s507wjc55llgh6sg2c2a9rc-flutter-engine-artifacts-3.7.12.drv
>   /nix/store/blbnc97qi1l3969y8rfr6v4b8b8z9sk8-source.drv
>   /nix/store/9xvdf179l0zrl85l83bqdak8byl0mdf4-dart-2.19.6.drv
>   /nix/store/6ar9khwcb5xsi0gifpk15gmhiy12adsr-flutter-3.7.12-unwrapped.drv
>   /nix/store/pnyjjl4d1lwhsrxg6mh52abdrm1zbvh4-flutter_immutable.drv
>   /nix/store/j6i4ibjckgrd5j6yl0i43yngl3i0hyrz-flutter-wrapped.drv
>   /nix/store/9wpmxw2f2fk5jy5gdnq0842b3jryd82c-flutter-wrapped-sdk-links.drv
> building '/nix/store/47m7nl28jk3nh47mf36b2807gx50g4sr-source.drv'...
> waiting for lock on '/nix/store/xbnq89ii1xfcvs4l5pdf583r1qf5lmhd-source'...
> building '/nix/store/blbnc97qi1l3969y8rfr6v4b8b8z9sk8-source.drv'...
> building '/nix/store/5mvrsprslvr0fg225xymjvyk15il8g45-source.drv'...
> building '/nix/store/y0kyyb5h946xpdw51k9g3lw3wf187yl9-source.drv'...
> building '/nix/store/yqd15cbi655kj6ysx5n1bjn1y71lydc5-source.drv'...
> building '/nix/store/yspn0q357wzbzmw2pa9l19194dyb8zj5-source.drv'...
> building '/nix/store/zaja9cgsn70ipdap8aknrb3w46fwdxnx-source.drv'...
> 
> trying https://storage.googleapis.com/flutter_infra_release/flutter/1a65d409c7a1438a34d21b60bf30a6fd5db59314/android-arm64-release/linux-x64.zip
> 
> trying https://storage.googleapis.com/flutter_infra_release/releases/stable/linux/flutter_linux_3.7.12-stable.tar.xz
> 
> trying https://storage.googleapis.com/flutter_infra_release/flutter/1a65d409c7a1438a34d21b60bf30a6fd5db59314/android-arm/artifacts.zip
> 
> trying https://storage.googleapis.com/dart-archive/channels/stable/release/2.19.6/sdk/dartsdk-linux-x64-release.zip
> 
> trying https://storage.googleapis.com/flutter_infra_release/flutter/1a65d409c7a1438a34d21b60bf30a6fd5db59314/linux-x64-profile/linux-x64-flutter-gtk.zip
> 
> trying https://storage.googleapis.com/flutter_infra_release/flutter/1a65d409c7a1438a34d21b60bf30a6fd5db59314/android-x86-jit-release/artifacts.zip
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>                                  Dload  Upload   Total   Spent    Left  Speed
>                                  Dload  Upload   Total   Spent    Left  Speed
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> 
> trying https://storage.googleapis.com/flutter_infra_release/flutter/1a65d409c7a1438a34d21b60bf30a6fd5db59314/android-x64-release/artifacts.zip
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> 100 9811k  100 9811k    0     0  20.1M      0 --:--:-- --:--:-- --:--:-- 20.1M
> unpacking source archive /build/linux-x64-flutter-gtk.zip
> error: hash mismatch in fixed-output derivation '/nix/store/5mvrsprslvr0fg225xymjvyk15il8g45-source.drv':
>          specified: sha256-RAsgupVF18IxLaP8tJ7XRQ8y/um46nlpA8fDITPwLqY=
>             got:    sha256-AnkLMPW3mwiXdiDz3Zo802QZRi+8EMCy4Mx2ODSbMOU=
> error: 1 dependencies of derivation '/nix/store/j6fgf61bpndxi0ivphdyh5hx6w117h61-flutter-artifact-linux-x64-profile-linux-x64-flutter-gtk-1a65d409c7a1438a34d21b60bf30a6fd5db59314.drv' failed to build
> error: 1 dependencies of derivation '/nix/store/498dcxxl3s507wjc55llgh6sg2c2a9rc-flutter-engine-artifacts-3.7.12.drv' failed to build
> error: 1 dependencies of derivation '/nix/store/6ar9khwcb5xsi0gifpk15gmhiy12adsr-flutter-3.7.12-unwrapped.drv' failed to build
> error: 1 dependencies of derivation '/nix/store/9wpmxw2f2fk5jy5gdnq0842b3jryd82c-flutter-wrapped-sdk-links.drv' failed to build
> ```

The artifact in question does seem to have been rebuilt recently. Many addresses and a few instructions in the binaries have changed.

By amazing coincidence the file was stable for days and changed just now...

hacker1024 commented:
> @hacker1024 you could pull request the fluffychat changes if you want now

https://github.com/NixOS/nixpkgs/pull/228282
